File: src/lib/worker/scheduler.ts
/**
 * Continuous re-enrichment scheduler. Once started (idempotently), it ticks
 * every TICK_MS milliseconds, runs Loop 1 in `delta` mode against every
 * source, and feeds any changed tables into Loop 2.
 *
 * Guard pattern: a single setInterval is stashed on globalThis so Next.js
 * hot reload doesn't spawn duplicates.
 *
 * The scheduler is the practical implementation of CLAUDE.md's "Loom is
 * always preparing" property — docs stay fresh as columns are added,
 * removed, or query patterns shift.
 */
import { catalogPool, sourcePool } from "../catalog/db";
import { runLoop1 } from "./loop1";
import { runLoop2ForTables } from "./loop2";
import { seedViewsForSource } from "./loop4-seed-views";
import { runAllWikiAgents } from "./wiki/runner";
import { runLoop3 } from "./loop3";
import { runEmbedBackfill } from "./embed-backfill";
import { audit } from "../catalog/queries";
import { log } from "../shared/log";

const TICK_MS = 60_000;        // 60s — slow enough to be cheap, fast enough to feel live
const MIN_GAP_PER_TABLE_MS = 90_000; // never re-enrich the same table more than ~once / 1.5 min
const WIKI_INTERVAL_MS = 24 * 60 * 60_000; // 24h — wiki regeneration is LLM-heavy (docs/code/domain agents call the doc-writer model)

declare global {
  var __loomSchedulerHandle: NodeJS.Timeout | undefined;
  var __loomSchedulerBusy: boolean | undefined;
  var __loomLastEnriched: Map<number, number> | undefined;
  var __loomLastWikiRunAt: number | undefined;
}

export function startScheduler(): { started: boolean; intervalMs: number } {
  if (globalThis.__loomSchedulerHandle) {
    return { started: false, intervalMs: TICK_MS };
  }
  globalThis.__loomLastEnriched ??= new Map();
  globalThis.__loomSchedulerHandle = setInterval(() => {
    void tickOnce().catch((e) => log.error("scheduler.tick_failed", { err: String(e) }));
  }, TICK_MS);
  log.info("scheduler.started", { intervalMs: TICK_MS });
  return { started: true, intervalMs: TICK_MS };
}

export function stopScheduler(): void {
  if (globalThis.__loomSchedulerHandle) {
    clearInterval(globalThis.__loomSchedulerHandle);
    globalThis.__loomSchedulerHandle = undefined;
    log.info("scheduler.stopped");
  }
}

async function tickOnce(): Promise<void> {
  if (globalThis.__loomSchedulerBusy) return;
  globalThis.__loomSchedulerBusy = true;
  try {
    // Active sources only (kind=postgres in v0.1 — other connectors are
    // recorded but not yet polled)
    const sources = await catalogPool.query<{ id: number; name: string; kind: string; conn_url: string }>(
      `SELECT id, name, kind, conn_url FROM sources WHERE kind = 'postgres'`,
    );
    for (const s of sources.rows) {
      try {
        const pool = sourcePool(s.conn_url);
        const { dirty } = await runLoop1(s.id, pool, "public", "delta");

        // Filter to tables that haven't been enriched too recently
        const now = Date.now();
        const eligible: number[] = [];
        const lastMap = globalThis.__loomLastEnriched!;
        for (const id of dirty) {
          const last = lastMap.get(id) ?? 0;
          if (now - last >= MIN_GAP_PER_TABLE_MS) {
            eligible.push(id);
            lastMap.set(id, now);
          }
        }
        if (eligible.length > 0) {
          await audit("system", "scheduler:reenrich", s.name, { tables: eligible.length });
          await runLoop2ForTables(s.conn_url, eligible);
        }

        // Loop 4 — proactive view seeding. Idempotent (no-ops if any
        // loom_views.* already exist for this source). Triggers naturally
        // the first time a source becomes "ready", and again when a fresh
        // demo:reset wipes views without wiping source data.
        try {
          await seedViewsForSource(s.id);
        } catch (e) {
          log.warn("scheduler.seed_failed", { source: s.name, err: String(e) });
        }
      } catch (e) {
        log.warn("scheduler.source_tick_failed", { source: s.name, err: String(e) });
      }
    }

    // Loop 3 — relationship discovery. Mines audit_log globally (not per
    // source), so it runs once per tick outside the source loop. Cheap when
    // there's nothing new: FK backfill is a no-op after the first run, and
    // observed-join mining only re-parses SQL from the last 24h.
    try {
      await runLoop3();
    } catch (e) {
      log.warn("scheduler.loop3_failed", { err: String(e) });
    }

    // Embedding backfill — embed any table doc or wiki page whose content
    // hash changed since the last embed. Batched (64 per tick) to keep API
    // spend bounded; full backfill of a 10K-table catalog takes ~160 ticks
    // (~3h) but is incremental, so it never blocks anything else.
    try {
      await runEmbedBackfill();
    } catch (e) {
      log.warn("scheduler.embed_backfill_failed", { err: String(e) });
    }

    // Wiki agents run at most once per day — docs/code/domain agents call
    // the doc-writer LLM, so even hash-gated they add up over many ticks.
    // On-demand paths bypass this (upload → docs agent, /api/wiki/code-sources
    // → code agent, /api/wiki/discover-domains → domains) so users can force
    // a refresh without waiting for the daily window.
    const now = Date.now();
    const lastWiki = globalThis.__loomLastWikiRunAt ?? 0;
    if (now - lastWiki >= WIKI_INTERVAL_MS) {
      globalThis.__loomLastWikiRunAt = now;
      try {
        await runAllWikiAgents();
      } catch (e) {
        log.warn("scheduler.wiki_failed", { err: String(e) });
      }
    }
  } finally {
    globalThis.__loomSchedulerBusy = false;
  }
}

File: backend/app/workers/scheduler.py
"""Continuous re-enrichment scheduler.

Slim port of src/lib/worker/scheduler.ts. Every TICK_SECONDS the scheduler
runs Loop 1 in delta mode for each active source, then Loop 2 over any
dirty tables. The TS app also runs Loop 3/4 + wiki agents in the same tick;
those ports come in later chunks.

Single asyncio.Task held in process-wide state — re-calling start() is a
no-op, matching the Node-side guard pattern.
"""
from __future__ import annotations

import asyncio
import logging
import time

from ..db import get_conn
from .loop1 import run_loop1
from .loop2 import run_loop2_for_tables

log = logging.getLogger(__name__)

TICK_SECONDS = 60
MIN_GAP_PER_TABLE_S = 90

_task: asyncio.Task[None] | None = None
_busy = False
_last_enriched: dict[int, float] = {}


def start_scheduler() -> bool:
    """Idempotent — returns True only on the call that actually started it."""
    global _task
    if _task is not None and not _task.done():
        return False
    _task = asyncio.create_task(_run_forever(), name="loom-scheduler")
    log.info("scheduler.started interval_s=%s", TICK_SECONDS)
    return True


async def stop_scheduler() -> None:
    global _task
    if _task is None:
        return
    _task.cancel()
    try:
        await _task
    except (asyncio.CancelledError, Exception):  # noqa: BLE001
        pass
    _task = None
    log.info("scheduler.stopped")


async def _run_forever() -> None:
    while True:
        try:
            await _tick()
        except Exception as e:  # noqa: BLE001
            log.error("scheduler.tick_failed err=%s", e)
        try:
            await asyncio.sleep(TICK_SECONDS)
        except asyncio.CancelledError:
            return


async def _tick() -> None:
    global _busy
    if _busy:
        return
    _busy = True
    try:
        sources = await _active_sources()
        for sid, name, conn_url in sources:
            try:
                result = await run_loop1(sid, conn_url, schema="public", mode="delta")
                now = time.monotonic()
                eligible: list[int] = []
                for tid in result.dirty:
                    last = _last_enriched.get(tid, 0.0)
                    if now - last >= MIN_GAP_PER_TABLE_S:
                        eligible.append(tid)
                        _last_enriched[tid] = now
                if eligible:
                    await run_loop2_for_tables(conn_url, eligible)
            except Exception as e:  # noqa: BLE001
                log.warning("scheduler.source_tick_failed source=%s err=%s", name, e)
    finally:
        _busy = False


async def _active_sources() -> list[tuple[int, str, str]]:
    async with get_conn() as conn:
        async with conn.cursor() as cur:
            await cur.execute(
                "SELECT id, name, conn_url FROM sources WHERE kind = 'postgres'"
            )
            rows = await cur.fetchall()
    return [(int(r[0]), str(r[1]), str(r[2])) for r in rows]
